pkg/local_object_storage/metabase/graveyard.go

```diff
@@ -9,70 +9,245 @@ import (
+type TombstonedObject struct {
+	addr *addressSDK.Address
+	tomb *addressSDK.Address
 }
 
 // Address returns buried object address.
-func (g *Grave) Address() *addressSDK.Address {
+func (g TombstonedObject) Address() *addressSDK.Address {
 	return g.addr
 }
 
-// GraveHandler is a Grave handling function.
-type GraveHandler func(*Grave) error
+// Tombstone returns address of a tombstone that
+// covers object.
+func (g TombstonedObject) Tombstone() *addressSDK.Address {
+	return g.tomb
+}
+
+// TombstonedHandler is a TombstonedObject handling function.
+type TombstonedHandler func(object TombstonedObject) error
+
+// GraveyardIterationPrm groups parameters of the graveyard
+// iteration process.
+type GraveyardIterationPrm struct {
+	h      TombstonedHandler
+	offset *addressSDK.Address
+}
+
+// SetHandler sets a handler that will be called on every
+// TombstonedObject.
+func (g *GraveyardIterationPrm) SetHandler(h TombstonedHandler) *GraveyardIterationPrm {
+	g.h = h
+	return g
+}
+
+// SetOffset sets an offset of the iteration operation.
+// The handler will be applied to the next after the
+// specified offset if any are left.
+//
+// Note: if offset is not found in db, iteration starts
+// from the element that WOULD BE the following after the
+// offset if offset was presented. That means that it is
+// safe to delete offset element and pass if to the
+// iteration once again: iteration would start from the
+// next element.
+//
+// Nil offset means start an integration from the beginning.
+func (g *GraveyardIterationPrm) SetOffset(offset *addressSDK.Address) *GraveyardIterationPrm {
+	g.offset = offset
+	return g
+}
 
 // IterateOverGraveyard iterates over all graves in DB.
 //
 // If h returns ErrInterruptIterator, nil returns immediately.
 // Returns other errors of h directly.
-func (db *DB) IterateOverGraveyard(h GraveHandler) error {
+func (db *DB) IterateOverGraveyard(p *GraveyardIterationPrm) error {
 	return db.boltDB.View(func(tx *bbolt.Tx) error {
-		return db.iterateOverGraveyard(tx, h)
+		return db.iterateDeletedObj(tx, graveyardHandler{p.h}, p.offset)
 	})
 }
 
-func (db *DB) iterateOverGraveyard(tx *bbolt.Tx, h GraveHandler) error {
-	// get graveyard bucket
-	bktGraveyard := tx.Bucket(graveyardBucketName)
-	if bktGraveyard == nil {
+type kvHandler interface {
+	handleKV(k, v []byte) error
+}
+
+type gcHandler struct {
+	h GarbageHandler
+}
+
+func (g gcHandler) handleKV(k, _ []byte) error {
+	o, err := garbageFromKV(k)
+	if err != nil {
+		return fmt.Errorf("could not parse garbage object: %w", err)
+	}
+
+	return g.h(o)
+}
+
+type graveyardHandler struct {
+	h TombstonedHandler
+}
+
+func (g graveyardHandler) handleKV(k, v []byte) error {
+	o, err := graveFromKV(k, v)
+	if err != nil {
+		return fmt.Errorf("could not parse grave: %w", err)
+	}
+
+	return g.h(o)
+}
+
+func (db *DB) iterateDeletedObj(tx *bbolt.Tx, h kvHandler, offset *addressSDK.Address) error {
+	var bkt *bbolt.Bucket
+	switch t := h.(type) {
```

Comment:
i tried that before and found that it is redundantly (IMO): 
1. handler contains info about bucket it could handle;
2. with an explicit parameter it could be possible to pass `garbage` bucket with `graveyardHandler`. that is totally incorrect but does not lead to panic or any other explicit errors => hard to debug

also, I think that sharing iteration logic is still good but I did not find any ways to do that without a type switch (btw, is a type switch that bad? maybe I do not know smth about it)

Reply:
No, type-switch is okay, just mentioned that it's needed only for bucket selection.